src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -213,73 +188,32 @@ public void Remove(object key)
 
             CheckDisposed();
             CacheEntry entry;
-            _entryLock.EnterReadLock();
-            try
-            {
-                if (_entries.TryGetValue(key, out entry))
-                {
-                    entry.SetExpired(EvictionReason.Removed);
-                }
-            }
-            finally
-            {
-                _entryLock.ExitReadLock();
-            }
-
-            if (entry != null)
+            if (_entries.TryRemove(key, out entry))
             {
-                // TODO: For efficiency consider processing these removals in batches.
-                RemoveEntry(entry);
+                entry.SetExpired(EvictionReason.Removed);
+                entry.InvokeEvictionCallbacks();
             }
 
             StartScanForExpiredItems();
         }
 
         private void RemoveEntry(CacheEntry entry)
         {
-            _entryLock.EnterWriteLock();
-            try
-            {
-                // Only remove it if someone hasn't modified it since our lookup
-                CacheEntry currentEntry;
-                if (_entries.TryGetValue(entry.Key, out currentEntry)
-                    && object.ReferenceEquals(currentEntry, entry))
-                {
-                    _entries.Remove(entry.Key);
-                }
-            }
-            finally
+            if (((ICollection<KeyValuePair<object, CacheEntry>>)_entries).Remove(new KeyValuePair<object, CacheEntry>(entry.Key, entry)))
```

Comment:
my eyes!!!
Store the `ICollection<KeyValuePair<object, CacheEntry>>` as another member field so you don't have to do this cast every time.